src/hashing.rs

```diff
@@ -0,0 +1,494 @@
+/// derivation is used in your application: for wallet derived keys, and communication derived keys. These might have
+/// the label "wallet-key" and "comms-key" respectively.
+///
+/// [`DomainSeparatedHasher`] is useful for more generic use-cases that aren't covered by the two primary use cases
+/// covered in this API (MAcs and key derivation).
+///
+/// ## Examples
+///
+/// Using a hash as an object ID, based on the fields of the object.
+///
+/// ```
+/// use sha2::Sha256;
+/// use tari_crypto::hashing::{DomainSeparatedHash, DomainSeparatedHasher, GenericHashDomain};
+/// use tari_utilities::hex::{to_hex, Hex};
+/// struct Card {
+///     name: &'static str,
+///     strength: u8,
+/// }
+///
+/// fn card_id(card: &Card) -> DomainSeparatedHash {
+///     DomainSeparatedHasher::<Sha256, GenericHashDomain>::new("card_id")
+///         .chain(card.name.as_bytes())
+///         .chain(&[card.strength])
+///         .finalize()
+/// }
+///
+/// let card = Card {
+///     name: "Rincewind",
+///     strength: 8,
+/// };
+///
+/// let id = card_id(&card);
+/// assert_eq!(id.domain_separation_tag(), "com.tari.generic.v1.card_id");
+/// assert_eq!(
+///     to_hex(id.as_ref()),
+///     "44fb39bfdd20c93ddf542e4b2d1f4b06448aa1fa2b9c4d138d8e7bbb19aa7c65"
+/// );
+/// ```
+///
+/// Calculating a signature challenge
+///
+/// ```
+/// use tari_crypto::{
+///     common::Blake256,
+///     hashing::{DomainSeparatedHash, DomainSeparatedHasher, GenericHashDomain},
+/// };
+/// use tari_utilities::hex::{to_hex, Hex};
+/// struct Card {
+///     name: &'static str,
+///     strength: u8,
+/// }
+///
+/// fn calculate_challenge(msg: &str) -> DomainSeparatedHash {
+///     DomainSeparatedHasher::<Blake256, GenericHashDomain>::new("schnorr_challenge")
+///         .chain(msg.as_bytes())
+///         .finalize()
+/// }
+///
+/// let challenge = calculate_challenge("All is well.");
+/// assert_eq!(
+///     challenge.domain_separation_tag(),
+///     "com.tari.generic.v1.schnorr_challenge"
+/// );
+/// assert_eq!(
+///     to_hex(challenge.as_ref()),
+///     "6cd8efe7d7f1673ed8cfc0ac67d6979eb50afdbf276adf5221caabfbfd01da8c"
+/// );
+/// ```
+pub struct DomainSeparatedHasher<D, M> {
+    inner: D,
+    label: String,
+    dst: PhantomData<M>,
+}
+
+impl<D: Digest, M: DomainSeparation> DomainSeparatedHasher<D, M> {
+    pub fn new<S>(label: S) -> Self
+    where S: AsRef<str> {
+        let inner = D::new();
+        let mut result = Self {
+            inner,
+            label: label.as_ref().to_string(),
+            dst: PhantomData::<M>::default(),
+        };
+        result.update(M::domain_separation_tag(&label).as_bytes());
+        result
+    }
+
+    /// Adds the data to the digest function by first appending the length of the data in the byte array, and then
+    /// supplying the data itself.
+    pub fn update(&mut self, data: impl AsRef<[u8]>) {
+        self.inner.update(data.as_ref().len().to_le_bytes());
```

Comment:
Perhaps change to either `u32` or `u64`, otherwise might be different on different targets depending on `usize`
```suggestion
        self.inner.update(data.as_ref().len().to_le_bytes());
```

Reply:
Good catch.
Should be `u64`